### PR TITLE
rework beam task workers for more flexibility

### PR DESCRIPTION
**Pull request by madeline-os**
The previous design was modeled largely after the task workers for groundhog which are constrained by the rigidity of that library. This design takes a new approach which says less about the shape of the table being used to store tasks and more about how to manipulate it with beam queries.

- [x] update the test suite
- [x] update the changelog

### CODE REVIEW

**Comment by madeline-os** [comment]
@cgibbard i think this probably should change so that it takes a connection pool and doesn't hold it through the whole worker task.

**Comment by ali-abrar** [comment]
@ymeister

**Comment by ymeister** [comment]
@ali-abrar don't completely follow what is expected of me here, so I branched [ym/split-beam-task-worker](https://github.com/obsidiansystems/rhyolite/tree/ym/split-beam-task-worker) into [ym/taskworker-update](https://github.com/obsidiansystems/rhyolite/tree/ym/taskworker-update) that is based on this branch, fixed all conflicts and remade a PR into this branch.

**Comment by madeline-os** [comment]
@cgibbard this is good to review now just waiting on CI

**Comment by Ericson2314** [comment]
Evidently https://github.com/obsidiansystems/rhyolite/pull/179 was rebased, because even after making github recompute the diff I still see too much diff (and duplicate commits causing that excess diff).

File: beam/task/types/src/Rhyolite/Task/Beam.hs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  -- ^ How to extract the payload from the row
+  , _task_checkedOutBy :: forall x. Lens' (table x) (C x (Maybe checkout))
+  -- ^ How the field which records a checkout is embedded within a row;
+  -- this allows both reading and writing.
+  , _task_hasRun :: forall x. Lens' (table x) (C x Bool)
+  -- ^ Which field indicates that the task result has been checked in.
+  , _task_result :: forall x. Lens' (table x) (result x)
+  -- ^ How the result data is embedded within a row.
+  -- Note that it has to be a Beamable type in its own right, so if the
+  -- result is a mere Columnar type it should be wrapped in a newtype.
```
Comment (plt-amy) [inline beam/task/types/src/Rhyolite/Task/Beam.hs:38]: `Beamable` and `Columnar` should probably be links?

File: beam/task/types/src/Rhyolite/Task/Beam.hs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+import Control.Lens
+
+import Database.Beam
+
+-- | The 'Task' type describes how to use a database table to check out tasks
+-- and report the results back.
+data Task be table payload checkout result = Task
+  { _task_filter :: forall s. table (QExpr be s) -> QExpr be s Bool
+  -- ^ How to filter unclaimed task rows
```
Comment (plt-amy) [inline beam/task/types/src/Rhyolite/Task/Beam.hs:27]: Could the wording here be improved to clarify what "unclaimed" means? AIUI it's something like:

```haskell
  { _task_filter :: forall s. table (QExpr be s) -> QExpr be s Bool
  -- ^ Filter for those rows which haven't been picked up by a worker
```

do I have that right?

File: beam/task/types/src/Rhyolite/Task/Beam.hs [hunk 3]
```diff
@@ -0,0 +1,39 @@
+{-# Language DeriveAnyClass #-}
+{-# Language DeriveGeneric #-}
+{-# Language FlexibleInstances #-}
+{-# Language FunctionalDependencies #-}
+{-# Language KindSignatures #-}
+{-# Language RankNTypes #-}
+{-# Language StandaloneDeriving #-}
+{-# Language TemplateHaskell #-}
+{-|
+Description : Contains definitions for tasks, and tables that have tasks.
```
Comment (plt-amy) [inline beam/task/types/src/Rhyolite/Task/Beam.hs:10]: What is "and tables that have tasks" referring to? I only see the `Task` type --- is GitHub not showing me the full diff? 🥴

**Review by plt-amy — APPROVED** [review]
I only have comments on the documentation 😔

File: beam/task/backend/src/Rhyolite/Task/Beam/Worker.hs [hunk 4]
```diff
@@ -26,80 +24,91 @@ import Database.Beam.Postgres.Syntax
...
+  -> DatabaseEntity be db (TableEntity table)
+  -- ^ The table whose rows represent tasks to be run
+  -> Task be table payload checkout result
+  -- ^ Description of how task data is embedded within the table
+  -- TODO: Serializable discipline here instead of 'Pg'
+  -> (PrimaryKey table Identity -> payload Identity -> Pg (m (Pg (result Identity))))
+  -- ^ The worker continuation is divided into 3 phases:
+  -- 1. A checkout action that is transaction safe (it may retry).
+  -- 2. A work action that is not transaction safe (it will not retry).
+  -- 3. A commit action that is transaction safe (it may retry).
```
[the remark this answers is not in the document]
Comment (madeline-os) [reply beam/task/backend/src/Rhyolite/Task/Beam/Worker.hs:61]: I will never intentionally make Haddock do anything in particular but since you noted it,,

**Comment by cgibbard** [comment]
This is why we need to get the CI system automatically hosting the
documentation it's already wasting its time building.

**Pull request merged**